drivers: esp32: I2C driver

lpereira opened the pull request:
This PR adds an implementation for the I2C device on ESP32 SoCs.  Sans possible bugs (not all configuration variations were tested), the driver is pretty much complete, supporting both master and slave modes, standard and fast speeds, and a bunch of ESP32-specific tunables.

lpereira committed:
drivers: esp32: I2C driver implementation

Supports both master and slave mode, standard and fast modes,
configurable timeouts, and a few other tunable settings.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
drivers: gpio: esp32: Use PIN_FUNC_GPIO definition from ESP-IDF

PINMUX_FUNC_A is set to 0, which coincides with the GPIO function in
many of the ESP32 pins.  Use PIN_FUNC_GPIO by default inside the
GPIO driver, however, so the correct function is always selected.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
drivers: gpio: esp32: Configure GPIO drive strength

Configuring an open drain driver is required by the I2C driver,
but the GPIO driver didn't support setting the drive strength.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	union dev_config config;
+	u16_t address;
+
+	struct k_sem fifo_sem;
```
These are wrong and will be fixed. (Macros to define the semaphore member should be used instead.)

tbursztyka commented on drivers/i2c/Kconfig.esp32:
```diff
@@ -0,0 +1,102 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config I2C_ESP32
```
menuconfig no?

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	if (status & (I2C_ARBITRATION_LOST_INT_RAW | I2C_ACK_ERR_INT_RAW)) {
+		return -EIO;
+	}
+	if (status & I2C_TIME_OUT_INT_RAW) {
```
add an empty line here

lpereira replied:
That would hurt readability: spaces are used to group things. In this case, this is grouping the reading of status bits. Adding a space would remove this visual clue.

pfalcon replied:
@tbursztyka would say: "Then use "else if" [and rely on compiler to not produce superfluous code, which is simple for if-else-return]". But I'm personally all for letting a programmer not to put a blank line where they don't want ;-).

tbursztyka replied:
please @pfalcon stop assuming what people would say.

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+			      struct i2c_msg msg)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	volatile struct i2c_esp32_cmd *cmd_base = (struct i2c_esp32_cmd *)config->regs.cmd_base;
```
too long line it seems

andyross replied:
Useful trick: make the cast a void*.  Only six characters for the type and guaranteed to assign cleanly to any pointer lvalue.  In a declaration like this it's where the type is already present it hurts nothing and improves readability.

lpereira replied:
I'll just change the config struct and make the cast where it's initialized.

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	if (data->config.bits.use_10_bit_addr) {
+		*config->regs.fifo = (addr >> 8) & I2C_FIFO_RDATA;
+	}
+	*cmd++ = (struct i2c_esp32_cmd) {
```
add an empty line. Basically, always break a line after a '}' unless it's another } etc..

Apply that everywhere (there are some places where it's needed)

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+
+	/* Send address */
+	*config->regs.fifo = addr & I2C_FIFO_RDATA;
+	if (data->config.bits.use_10_bit_addr) {
```
You can coalesce an assignment and an if/for/while if the condition in those test the previous assignment. 
Here it's not the case so add an empty line.

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	u32_t opcode : 3;
+	u32_t : 17;
+	u32_t done : 1;
+};
```
__packed? (though it's one u32_t in the end, maybe it's not needed)

andrewboie replied:
shouldn't need __packed here

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+
+	const union dev_config default_config;
+};
+
```
For all the above declarations, it's usually nicer to move them in a dedicated header file, where you can document them as well.

andyross commented on drivers/i2c/Kconfig:
```diff
@@ -210,7 +211,7 @@ config I2C_0_DEFAULT_CFG
 config I2C_0_IRQ_PRI
 	int "Port 0 interrupt priority"
-	depends on I2C_0 && !HAS_DTS_I2C
+	depends on I2C_0 && !HAS_DTS_I2C && !I2C_ESP32
```
Not understsanding this one.  Can't you just turn off HAS_DTS_I2C for the esp32 board if the board configuration doesn't include the I2C controller in the devicetree configuration?  Isn't that the point of HAS_DTS_I2C in the first place?

mnkp replied:
The I2C device drivers are being converted to use DTS: #755. Since this is a new driver it would be best if it already used DTS. Adding DTS support is not much work. All these Kconfig dependencies will become much simpler.

andyross commented on drivers/i2c/Kconfig:
```diff
@@ -236,13 +237,14 @@ config I2C_1_DEFAULT_CFG
 config I2C_1_IRQ_PRI
 	int "Port 1 interrupt priority"
-	depends on I2C_1 && !HAS_DTS_I2C
+	depends on I2C_1 && !HAS_DTS_I2C && !I2C_ESP32
```
Same comment

andyross commented on drivers/i2c/Kconfig:
```diff
@@ -236,13 +237,14 @@ config I2C_1_DEFAULT_CFG
 config I2C_2
 	bool "Enable I2C Port 2"
 	default n
+	depends on !I2C_ESP32
```
Likewise

andyross commented on drivers/i2c/Kconfig:
```diff
@@ -269,6 +271,7 @@ config I2C_2_IRQ_PRI
 config I2C_3
 	bool "Enable I2C Port 3"
 	default n
+	depends on !I2C_ESP32
```
Same

andyross commented on drivers/i2c/Kconfig.esp32:
```diff
@@ -0,0 +1,102 @@
+
+if I2C_0
+
+config I2C_0_DEFAULT_CFG
```
Seems obscure.  I guess this is a bitmask of a config register?  Can't you split out the fields you want configurable and compute the bitmask in a compile-time C expression?  Needs docs anyway.  No one will know what to do with this at the kconfig level.

lpereira replied:
This is just how our I2C subsystem is configured.  I don't like this as well, and would prefer separate settings, but it's something to be fixed in another pull request.

andyross replied:
Got it.  Yeah, your'e right.  Not your problem to fix in this patch.

mnkp replied:
This issue has already been fixed #1081. Drivers which use DTS can define start-up configuration in a clean way.

andyross commented on drivers/i2c/Kconfig.esp32:
```diff
@@ -0,0 +1,102 @@
+	default n
+
+config I2C_ESP32_0_IRQ
+	int "Port 0 IRQ line"
```
Is this actually configurable?  Almost certain everything on an xtensa SoC is hardwired to specific IRQs.  Or maybe different ESP-32 variants use different lines?  If it can't change across hardware we actually know about, this should probably be a preprocessor symbol and not a kconfig.

lpereira replied:
Yes, it's configurable. There are many IRQ lines that can be routed to peripherals; there's a call to a ROM function that controls the interrupt matrix to set this up.

andyross commented on drivers/i2c/Kconfig.esp32:
```diff
@@ -0,0 +1,102 @@
+
+config I2C_ESP32_0_SCL_PIN
+	int "Port 0 SCL pin"
+	default 2
```
Likewise, unless the internal mux allows for moving the pins around this seems like something that shouldn't be in kconfig.

lpereira replied:
Any GPIO pin can be SCL or SDA.

andyross commented on drivers/i2c/Kconfig.sbcon:
```diff
@@ -6,7 +6,7 @@
 menuconfig I2C_SBCON
 	bool "I2C driver for ARM's SBCon two-wire serial bus interface"
-	depends on I2C
+	depends on I2C && ARM
```
Why add ARM?  Just don't include it on the board config, right?

lpereira replied:
Oops. This should be in another pull request.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+struct i2c_esp32_config {
+	const struct {
+		volatile u32_t *ctr;
+		volatile u32_t *fifo_cfg;
```
This scheme seems clumsy.  You're building and storing a distinct pointer to every register?  Why not have a "struct esp32_i2c_regs" structure which specifies a single struct with the register array and keep a **single** volatile pointer to that?

Or maybe you're using "pointer to volatile u32_t" as the **type** of the register (which is also wrong: volatile goes on the thing pointed to not the pointer itself)?

Basically: this is a super-giant config struct, and I dont' understand why.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+};
+
+static void i2c_esp32_set_device_enabled(const struct i2c_esp32_config *config,
+					 bool setting)
```
Pedantry: "setting" should be named "is_enabled" or something similar so we know which direction the boolean goes.

Also: given that you have to test "setting" in the implementation, this is going to involve more code and run slower than simply having two functions named i2c_esp32_enable_device() and i2c_esp32_disable_device().

lpereira replied:
Being such a short static function, where the `setting` argument is always passed as a constant value, the compiler will most likely inline it and eliminate dead code.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	*config->regs.scl_stop_h_time =
+		(period / 2) << I2C_SCL_STOP_HOLD_TIME_S;
+	*config->regs.scl_stop_s_time =
+		(period / 2) << I2C_SCL_STOP_SETUP_TIME_S;
```
If those are shift/exponent values and not linear multipliers, should probably have more than just a _S appended to cue the reader.

lpereira replied:
These are defined by esp-idf.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	union dev_config dev_config = { .raw = dev_config_raw };
+	int key = irq_lock();
+	u32_t v;
+	int r;
```
Style: probably cleaner to put the pure declarations for "v" and "r" above the declaration for "key" which has an obvious side effect.

lpereira replied:
If I change this, @tbursztyka will come in and say that, for style reasons, variables declarations should happen from the largest to the smallest length. :)

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+
+	v = *config->regs.ctr;
+
+	/* MSB or LSB first is configurable for both TX and RX */
```
So... do we care?  Gotta be honest I didn't know configurable bit order was even a thing for I2C.  Are there any peripherals that actually care or want this?  Maybe drop this feature, or at least make it a sanely defaulted kconfig that most users don't ever touch.

lpereira replied:
I've abused these things in the past.  Once I wanted to control an array of WS2812B LEDs using Linux, but it's very difficult to obtain precise timings, specially when you have hundreds of daisy-chained LEDs.  By using the [SPI controller and framing the bits just right](https://hackaday.com/2014/09/10/driving-ws2812b-pixels-with-dma-based-spi/), it's possible to convince it can speak the WS2812B protocol.  And sometimes, doing things like this requires doing things that are not standard.

Not, to answer your question: I don't have a clue if we care.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	}
+
+	if (dev_config.bits.is_master_device) {
+		/* Act as master. */
```
Style: a test on "is_master_device" predicating a set of a "MS_MODE" bit is documentation enough.  You don't need a comment to tell us that this is setting the master mode.

pfalcon replied:
Given there's `/* Act as slave. */` below, I guess `/* Act as master. */` was added for symmetry and consistency.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+
+	i2c_esp32_set_device_enabled(config, true);
+
+	v = *config->regs.ctr;
```
It looks like you're writing every field, or nearly every field in this register.  Why start by reading out the original contents?  Seems like unless there is something there that is undocumented we'd be more robust by affirmatively setting every bit.  Would also shrink code size for all the expressions below that turn bits off when they could be left at an initial zero state.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+		/* Before setting up FIFO and interrupts, stop transmission */
+		*config->regs.ctr &= ~I2C_TRANS_START;
+
+		/* Byte after address isn't the offset address in slave RAM */
```
You mean "is", not "isn't", right?  Or... if not that, what is it?  Confusing comment, though maybe it makes sense if I were to read the docs.

lpereira replied:
It's correct, but maybe this would benefit from better wording. The documentation has this as: 

> **I2C_FIFO_ADDR_CFG_EN** When this bit is set to 1, the byte received after the I2C address byte
> represents the offset address in the I2C Slave RAM

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	v |= I2C_CLK_EN;
+
+	/* Sample data when SCL is high */
+	v &= ~I2C_SAMPLE_SCL_LEVEL;
```
Amused quip: the device supports two bit orders and an inverted clock?  Sounds to me like maybe the hardware designers just didn't know which convention was right so figured they'd implement both just to be safe...

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+		break;
+	case I2C_SPEED_FAST:
+		i2c_esp32_configure_frequency(config, 400000);
+		break;
```
Contra the extra stuff I flagged in kconfig, this seems to be a spot where you're hard-coding values that should probably be configurable.  Hardware in the wild tends to support wide varieties of speeds, and I'm told people like to tune the bus rates on real devices for things like latency and power.  Probably make these two config variables unless there's some reason the hardware needs exactly 100/400kHz.

mnkp replied:
That's a very good point. By specifying bus speed value as a number we would not only be more flexible but also reduce the code size. This whole case structure could be gone. That said, our current I2C API supports only fixed bus speed values so first we need to change the API. I'll open a Jira issue.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+
+static inline void i2c_esp32_ack_intrs(const struct i2c_esp32_config *config)
+{
+	/* Acknowledge all interrupts */
```
Is this known to be safe?  Often documentation puts stuff like "reserved" in the unused bit fields of an interrupt register and expects you not to ack them spuriously.  Maybe safer to write back the mask of known valid (and handled) interrupts.

lpereira replied:
These are all I2C interrupts, so it's safe.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	int counter = 0;
+	int r;
+
+	i2c_esp32_ack_intrs(config);
```
This looks like a race.  Are you absolutely, positively sure that before the entry to this function it's not possible even in the abstract for an interrupt to be posted to the device from any source anywhere?  If not, this is going to potentially ACK it before the interrupt is delivered to the CPU.

What's the purpose anyway?  We should be ACKing interrupts inside interrupt handlers, after having handled them.  What does this do?   If you want to be sure the hardware is at a known state, maybe call it "reset" instead?

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+		}
+	}
+
+	/* Wait for I2C bus to finish its business */
```
Put this at the start of the function, not the end.  That way whatever framing cleanup is happening in the hardware can parallelize with the software handling of the command result.

lpereira replied:
I can't find the context for this now (GitHub doesn't seem to give context for "outdated" hunks), but this is related to the function call to clear the FIFO, I'm not sure if I follow you:

* The FIFO is cleared in the beginning of the function, then the address is written there
* It has to be cleared for every block of transmission, and we need to wait for it to finish
* If we leave each loop iteration without clearing the FIFO (say, if the FIFO is cleared before commands are queued), this could be a race condition if a transmission is started right after the one that just finished
* Clearing the FIFO is just writing to two hardware registers, it should be pretty quick

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	/* Start transaction */
+	if (msg.flags & I2C_MSG_RESTART) {
+		*cmd++ = (struct i2c_esp32_cmd) {
+			.opcode = I2C_ESP32_OP_RSTART
```
I'm sure someone is going to complain about this expression (postincrement pointer assignment of a casted structure literal using GCC field syntax).  So let me preemptively say I love it and it should stay.

andyross reviewed (changes requested):
Not quite done yet.  Figure I should post what I have before I lose it.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	};
+
+	for (cmd_after_write = cmd; msg.len; cmd = cmd_base) {
+		volatile struct i2c_esp32_cmd *wait_cmd = NULL;
```
The pointers here are RAM (the command being set up for transmission), not mapped memory on the device, right?  I don't think the volatile is doing anything useful for you here.  We don't care about side effects here, gcc can see all that anyway.  And remember "volatile" doesn't have anything to do with cache consistency.  If that's what you're worried about (in the abstract) you need to include a memory barrier regardless.

Basically remove the "volatile" or document in code the subtle reason it's required.

lpereira replied:
No, these are pointers to peripheral registers.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+		u32_t to_read = max(I2C_ESP32_BUFFER_SIZE, msg.len - 1);
+
+		/* Might be the last byte, in which case, `to_read` will
+		 * be 0 here.  See comment below.
```
Seems like another good clue that maybe the loop iteration design needs rework so we don't need comments like this. :)

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+		}
+
+		/* I2C master won't acknowledge the last byte read from the
+		 * slave device.  Divide the read command in two segments as
```
It won't... what now?  Sigh...

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+		}
+		msg.len -= to_read;
+
+		i2c_esp32_reset_fifo(config);
```
Similar to the earlier point: it's generally better to put "hardware reset/cleanup" stuff at the **start** of a command instead of the end.  The software layer usually wants its output as soon as possible and is willing to pay some setup latency.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	u32_t addr_len = 1 + data->config.bits.use_10_bit_addr;
+	bool first = true;
+
+	for (msg.len += addr_len; msg.len; ) {
```
Much cleaner loop than the read one.  Still might be more simply expressed as a while by taking that "include the address length" bit out and making it a separate expression.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+
+	for (msg.len += addr_len; msg.len; ) {
+		volatile struct i2c_esp32_cmd *cmd =
+			(struct i2c_esp32_cmd *)config->regs.cmd_base;
```
Same comment about volatile.  I don't think it's doing what you think it is.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+				I2C_ESP32_OP_END : I2C_ESP32_OP_STOP,
+		};
+
+		/* Ensure the remaining I2C commands are cleared */
```
Why?  Isn't there a "command count" register or something similar?  Or the hardware really wants to read out zero-valued commands as "noops"?

lpereira replied:
This is done by the FreeRTOS driver as well.  I don't know if this is strictly necessary (i.e. the controller will stop when an END or a STOP opcode is found).

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+		.scl_stop_s_time = (u32_t *)I2C_SCL_STOP_SETUP_REG(0),
+		.sda_hold_time = (u32_t *)I2C_SDA_HOLD_REG(0),
+		.sda_sample_time = (u32_t *)I2C_SDA_SAMPLE_REG(0),
+	},
```
Yeah, this is the 76-byte mess I was complaining about when I saw the struct definition.  I just don't see the point, why not make sure there's a bus number available everywhere in code where you have a handle to the I2C device and just use those macros directly?  I don't see that this adds anything but extra code and rodata, for no benefit.

lpereira replied:
I did the same thing in the GPIO driver, but things got out of hand for I2C, yeah.  Will store just `0` or `1` and use these registers directly.

mnkp replied:
CMSIS standard used by the ARM processors solves this issue nicely using define macros. You can have a look at the ext/hal/atmel/asf/sam/include/sam3x/component/twi.h file and then ext/hal/atmel/asf/sam/include/sam3x/sam3x8e.h where TWI0 and TWI1 (I2C modules on Atmel SoCs) are defined as 
```
#define TWI0       ((Twi    *)0x4008C000U) /**< \brief (TWI0      ) Base Address */
#define TWI1       ((Twi    *)0x40090000U) /**< \brief (TWI1      ) Base Address */
```
You can check how this is used in the code by looking at the Atmel SAM I2C driver at drivers/i2c/twihs_sam.c. Maybe you find it helpful.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+	intr_matrix_set(0, config->irq.source, config->irq.line);
+
+	/* These if statements below are necessary as IRQ_CONNECT() requires
+	 * a constant argument.
```
Not understanding the comment.  The expression "CONFIG_I2C_ESP32_0_IRQ" is going to be a constant for sure.  The variable in the config struct won't, but that's not what you're using.

Also: as mentioned before I really don't think these IRQs are actually configurable in any meaningful way.  I'm pretty sure they're constant at least within an SoC, which can be tested at compile time and removed from kconfig.

lpereira replied:
This function is called twice: one for each I2C controller. Different lines are assigned for them.

The IRQ lines are configurable: what's fixed are how they're triggered (level, edge, etc), their priorities, and category (peripheral or internal).  If it's a peripheral interrupt line, sources can be routed (intr_matrix_set() call above).

andyross replied:
Ah, got it.  You want each "called" only once.

The dirty secret here though is that IRQ_CONNECT() doesn't actually produce any runtime code.  It's perfectly valid for you to set up both interrupts in this one function and "call" them twice at runtime.  Or do it from a different global init spot, etc...  All that needs to happen is that the compiler expands the macro only once.

Though thinking about it, the compiler will just eliminate the tests anyway as all its sees are empty blocks following the two tests.

Honestly IRQ_CONNECT() needs some rethinking as a framework.

mnkp replied:
This can be implemented cleanly by defining an extra per-instance IRQ configuration function as described in Zephyr Device Drivers Documentation section [Single Driver, Multiple Instances](https://www.zephyrproject.org/doc/devices/drivers/drivers.html#single-driver-multiple-instances). The documentation calls this function 'config_func'. I find the name pretty confusing however, it is too general and there are already many config-s in the driver. In the device drivers I wrote I use always the name 'irq_config'.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,781 @@
+			   pull_up ? GPIO_PUD_PULL_UP : GPIO_PUD_NORMAL);
+	gpio_pin_write(gpio, pin, default_level);
+
+	gpio_matrix_out(pin, matrix_out, 0, 0);
```
Given the naming is ambiguous and generic: probably worth a comment explaining that these calls are to the ESP-32 ROM-resident library layer and not part of this driver or Zephyr's gpio stack.

Or maybe we want to do some abstracting of our own and rename these symbols for our own use with esp32_rom_* prefixes or something.  Not really relevant to this patch.

lpereira committed:
esp32: Prefix ROM routines with `esp32_rom_`

Also provide their prototypes in `soc.h`.  This should help
readability, since some ROM functions, with their names as provided by
Espressif, have sometimes the same prefix as Zephyr APIs.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
docs: drivers: Fix syntax errors in "Single Driver, Multiple Instances"

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

dbkinder reviewed (dismissed):
+1 for doc change

lpereira commented:
@andyross @mnkp @tbursztyka Here's a new version -- not every suggested thing has been changed in this version, but it's a lot cleaner now.

galak reviewed (dismissed):
Please update to remove use of 'union dev_config' as per:

commit 83643a2266b78f0b39c954a1f14bd68c4d720332
Author: Kumar Gala <kumar.gala@linaro.org>
Date:   Wed Sep 13 11:31:00 2017 -0500

    drivers: i2c: remove usage of bitfield access for cfg

mnkp commented on drivers/i2c/Kconfig.esp32:
```diff
@@ -0,0 +1,102 @@
+if I2C_ESP32
+
+config I2C_ESP32_TIMEOUT
+	int "I2C timeout to receive a data bit in APB clock cycles"
```
Shouldn't we use 'us' as units?

lpereira commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,725 @@
+	/* Use open-drain for clock and data pins */
+	v |= (I2C_SCL_FORCE_OUT | I2C_SDA_FORCE_OUT);
+	v |= I2C_CLK_EN;
+	sys_write32(I2C_CTR_REG(config->index), v);
```
I don't see a way out here.

mnkp replied:
You are right. It's `v |= `, the other style does not work here. And converting all the places where you pass flags directly to sys_write32() to use this style is not worth it.

lpereira commented on boards/xtensa/esp32/esp32_defconfig:
```diff
@@ -28,3 +28,13 @@ CONFIG_RANDOM_ESP32_RNG=y
+CONFIG_I2C_1_IRQ_PRI=1
+CONFIG_I2C_2=n
+CONFIG_I2C_3=n
+CONFIG_HAS_DTS_I2C=n
```
I'm with you here, but I'd like to see the Xtensa port be converted to DTS before having parts of it defined in Device Tree and parts of it in Kconfig.

mnkp replied:
Good point, you're right.

lpereira commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+}
+
+static int i2c_esp32_write_msg(struct device *dev, u16_t addr,
+			       struct i2c_msg msg)
```
There's a possibility it would, yes.  However, both `read` and `write` functions modifies the `msg` parameter (by advancing the `buf` pointer and modifying the `len` variable).  So the copy is safer.

lpereira commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+{
+	struct i2c_esp32_data *data = dev->driver_data;
+	int ret = 0;
+	u8_t i;
```
You would be right, but `num_msgs` is an `u8_t`, and it's part of the Zephyr API.  Using an `int` here would mean that extra instructions would have to be generated, not the other way around.

mnkp replied:
I still think I'm right. Casting u8_t to int is not a problem / comes for free, casting int to u8_t requires extra code. Compiler will use 32-bit register variable to do arithmetic, even if it is declared as u8_t.

lpereira commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+	k_sem_take(&data->transfer_sem, K_FOREVER);
+
+	/* Mask out unused address bits, and make room for R/W bit */
+	addr &= BIT_MASK(data->dev_config & I2C_ADDR_10_BITS ? 10 : 7);
```
We don't do much error handling in Zephyr.  I prefer not to add another branch here.

andrewboie replied:
I concur with Leandro here.
At most, put an assertion.

lpereira commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+	for (; msg.len; cmd = (void *)I2C_COMD0_REG(config->index)) {
+		volatile struct i2c_esp32_cmd *wait_cmd = cmd;
+		u32_t to_send = min(I2C_ESP32_BUFFER_SIZE, msg.len);
+		u32_t i;
```
I can change this, but is this really necessary? I guess we're at the point in the review that only nits are being picked.

andrewboie replied:
this is not necessary, let's stay focused on functionality / correctness / performance

mnkp replied:
> I guess we're at the point in the review that only nits are being picked.

Definitely. For me an extra spin is worth it if it improves overall readability of the code. Code tends to be read a lot. In any case my recent comments certainly fall into a category 'Style whine' as @andyross calls it.

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+}
+
+static int i2c_esp32_configure_frequency(const struct i2c_esp32_config *config,
+					 u32_t speed)
```
Style whine: function name says "frequency", parameter name is "speed", and what it actually is is an enumerant.  Trying to pass this thing a frequency number in khz or whatever is going to generate a giant buffer overrun.   Really the i2c layer should have defined these as an enum to begin with I  guess.  Maybe rename the parameter "speed_enum" or something for clarity if you do another spin?

andyross commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+	set_mask32(period << I2C_SCL_HIGH_PERIOD_S,
+		   I2C_SCL_HIGH_PERIOD_REG(config->index));
+
+	period /= 2; /* Set hold and setup times to 1/2th of period */
```
Do these particular settings come from a design guide or something somewhere?  Might be worth citing it so people understand where the magic comes from.

andyross reviewed (dismissed):
Can't see anything worth much complaint anymore

tbursztyka commented on drivers/i2c/Kconfig.esp32:
```diff
@@ -0,0 +1,102 @@
+
+if I2C_ESP32
+
+config I2C_ESP32_TIMEOUT
```
Looks like to me this timeout is very specific to this driver.

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+	if (!device_name) {
+		return -EINVAL;
+	}
+	gpio = device_get_binding(device_name);
```
add an empty line before

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+		if (dev_config & I2C_ADDR_10_BITS) {
+			addr |= I2C_ADDR_10BIT_EN;
+		}
+		sys_write32(addr << I2C_SLAVE_ADDR_S,
```
add an empty line before

tbursztyka commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -0,0 +1,713 @@
+
+	/* Even if irq_enable() is called on config->irq.line, disable
+	 * interrupt sources in the I2C controller.
+	 */
```
Is there a specific esp32 behavior that requires the driver to specifically disable the irq to re-enable them at configure time?
Usually, irq_enable() is ran after IRQ_CONNECT() and disabling the interrupts from the controller is sufficient to silence it, thus why I am asking.

lpereira replied:
ESP32 has two CPUs, and both can fire interrupts.  While the second CPU is currently halted, there are plans to support SMP in the near future.
